src/cockpit/samples.py

```diff
@@ -1,65 +1,300 @@
+        SampleDescription('cgroup.memory.limit', 'bytes', 'instant', True),
+        SampleDescription('cgroup.memory.sw-usage', 'bytes', 'instant', True),
+        SampleDescription('cgroup.memory.sw-limit', 'bytes', 'instant', True),
+        SampleDescription('cgroup.cpu.usage', 'millisec', 'counter', True),
+        SampleDescription('cgroup.cpu.shares', 'count', 'instant', True),
+    ]
+
+    cgroups_v2: Optional[bool] = None
+    cgroups_v2_path = '/sys/fs/cgroup/'
+
+    def read_cgroup_keyed_stat(self, samples, path, cgroup, name, key):
+        with open(path) as stat:
+            for line in stat:
+                if not line.startswith(key):
+                    continue
+
+                value = int(line.split()[-1])
+                if sys.maxsize > value > 0:
+                    samples[name][cgroup] = value / 1000
+
+    def read_cgroup_integer_stat(self, samples, path, cgroup, name):
+        # Not every stat is available, such as cpu.weight
+        try:
+            with open(path) as stat:
+                # Some samples such as "memory.max" contains "max" when there is a no limit
+                try:
+                    value = int(stat.read().strip())
+                except ValueError:
+                    return
+
+                if sys.maxsize > value > 0:
+                    samples[name][cgroup] = value
+        except FileNotFoundError:
+            pass
+
+    def sample(self, samples):
+        # TODO: Cgroups v1 support
+        if self.cgroups_v2 is None:
+            self.cgroups_v2 = os.path.exists('/sys/fs/cgroup/cgroup.controllers')
+
+        if self.cgroups_v2_path:
+            for path, _, _ in os.walk(self.cgroups_v2_path):
+                cgroup = path.replace(self.cgroups_v2_path, '')  # TODO: Pathlib?
+
+                if not cgroup:
+                    continue
+
+                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.current'), cgroup, 'cgroup.memory.usage')
+                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.max'), cgroup, 'cgroup.memory.limit')
+                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.swap.current'), cgroup, 'cgroup.memory.sw-usage')
+                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.swap.max'), cgroup, 'cgroup.memory.sw-limit')
+                self.read_cgroup_integer_stat(samples, os.path.join(path, 'cpu.weight'), cgroup, 'cgroup.cpu.shares')
+                self.read_cgroup_keyed_stat(samples, os.path.join(path, 'cpu.stat'), cgroup, 'cgroup.cpu.usage', 'usage_usec')
+
+
+class NetworkSampler(BaseSampler):
+    descriptions = [
+        SampleDescription('network.interface.tx', 'bytes', 'counter', True),
+        SampleDescription('network.interface.rx', 'bytes', 'counter', True),
+    ]
+
+    def sample(self, samples):
+        with open("/proc/net/dev") as network_samples:
+            for line in network_samples:
+                lines = line.split()
```

Comment:
huh?

You iterate the file line-by-line, but then you split each line up and store it into a variable called "lines".  I think either this isn't doing what you think, or I missed something.